### PR TITLE
Dashboard API

### PR DESCRIPTION
**Pull request by TheMightyRaider**
- Created a new table for storing streak details.
- Added a endpoint for retrieving the dashboard data.
- Added logic for handling streak updates.
- Updated **/userlog** endpoint to provide **mode** details
- Updated PractiseLog model.
- Fixed the login error, i.e user can login from multiple system at the same time.

### CODE REVIEW

File: typebackend/serializers.py [hunk 1]
```diff
@@ -18,12 +18,12 @@ def create(self,validated_data):
 class PractiseLogSerializer(serializers.ModelSerializer):
     class Meta:
         model=PractiseLog
-        fields=["para","wpm","taken_at","correct_words","wrong_words","total_words","accuracy"]
+        fields=["para","wpm","taken_at","correct_words","wrong_words","total_words","accuracy","mode"]
```
Comment (digi0ps) [inline typebackend/serializers.py:21]: Why note "__all__" ?

Comment (digi0ps) [reply typebackend/serializers.py:21]: `__all__`

File: typebackend/serializers.py [hunk 2]
```diff
@@ -18,12 +18,12 @@ def create(self,validated_data):
...
         extra_kwargs = {'para': {'write_only': True}}
 
     def create(self,validated_data):
         user=self.context['request'].user
-        result=user.practiselog_set.create(para=validated_data['para'],wpm=validated_data['wpm'],correct_words=validated_data['correct_words'],wrong_words=validated_data['wrong_words'],total_words=validated_data['total_words'],accuracy=validated_data['accuracy'],taken_at=validated_data['taken_at'])
+        result=user.practiselog_set.create(para=validated_data['para'],wpm=validated_data['wpm'],correct_words=validated_data['correct_words'],wrong_words=validated_data['wrong_words'],total_words=validated_data['total_words'],accuracy=validated_data['accuracy'],taken_at=validated_data['taken_at'],mode=validated_data['mode'])
```
Comment (digi0ps) [inline typebackend/serializers.py:26]: I don't think you need to do this, DRF automatically takes care if all the variable names are the same as the model field names.

File: typebackend/typingviews.py [hunk 3]
```diff
@@ -3,26 +3,64 @@
...
+        last_typed=new_entry-datetime.timedelta(1)
+
+    # Finding the difference in days between the last log and the new log.
+    days_between_recent_and_lastlog=(new_entry-last_typed).days
+    data=DashboardData.objects.get(user=user)
+
+    # Getting the existing value of the mode [practise/arcade/race] specified in the user object from the query and updating it.
+    count=getattr(data,mode)+1
+    #Setting the updated value of the mode count
+    setattr(data,mode,count)
```
Comment (digi0ps) [inline typebackend/typingviews.py:33]: Why not:
```
data[mode] = data[mode] + 1
```

File: typebackend/typingviews.py [hunk 4]
```diff
@@ -3,26 +3,64 @@
...
+
+    # Checking for consecutive streak from the database, i.e if the difference between previous log and new log is 1, then the user has been typing consecutively
+    if days_between_recent_and_lastlog==1:
+        data.streak=data.streak+1
+        data.total_streak=1 if data.total_streak==0 else 0 # Incase of new user
+        data.longest_streak=data.streak if data.streak>data.longest_streak else data.longest_streak
```
Comment (digi0ps) [inline typebackend/typingviews.py:39]: Set total_streak default as 0 in models.py.
Make both to a if's without `else`

Comment (digi0ps) [reply typebackend/typingviews.py:39]: You don't actually need L38 if you set default

File: typebackend/typingviews.py [hunk 5]
```diff
@@ -3,26 +3,64 @@
...
     permission_classes=[IsAuthenticated]
 
     def post(self,request):
-        serializers=PractiseLogSerializer(data=request.data,context={'request':request})
-        
-        if serializers.is_valid():  
-            user=serializers.save()
-            return Response({'success':True})
-        else:
+        if request.data['mode']=='practise' or request.data['mode']=='race' or request.data['mode']=='arcade':
```
Comment (digi0ps) [inline typebackend/typingviews.py:52]: Why this condition?

File: typebackend/typingviews.py [hunk 6]
```diff
@@ -3,26 +3,64 @@
...
+            serializers=PractiseLogSerializer(data=request.data,context={'request':request})
+            if serializers.is_valid():
+                user=request.user
+                typed_at=request.data['taken_at']
+                mode=request.data['mode']
+                # Streak counter update 
+                create_or_update_streak(user,typed_at,mode)
```
Comment (digi0ps) [inline typebackend/typingviews.py:59]: Actually this whole `create_or_update_streak` can be moved as a Django signal.
Read about on_create signal. It doesn't block the request thread.

File: typebackend/typingviews.py [hunk 7]
```diff
@@ -91,4 +129,24 @@ def get(self,request):
...
+    permission_classes=[IsAuthenticated]
+
+    def get(self,request):
+        # { avg wpm, avg accuracy, user since, streak, Inactive days, longest streak, Total stresks, mode count }
+        user_id=request.user.id
+
+        dashboard_data={}
+        dashboard_data['user_since']=str(User.objects.get(id=user_id).date_joined)[0:10]
+        total_log=PractiseLog.objects.filter(user=user_id).count()
+        wpm_and_accuracy=PractiseLog.objects.filter(user_id=user_id).aggregate(wpm=Sum('wpm')/total_log,accuracy=Sum('accuracy')/total_log)
```
Comment (digi0ps) [inline typebackend/typingviews.py:144]: Add this in Dashboard model too. These are persistent and expensive to calculate data.

File: typebackend/urls.py [hunk 8]
```diff
@@ -9,5 +9,6 @@
     path('userlog',typingviews.PostSpeed.as_view(),name='userlog'),
     path('para',typingviews.Paradetails.as_view(),name='para'),
     path('internal/racetrack/para',typingviews.RaceTrack.as_view(),name='racetrack'),
-    re_path(r'^getuserlog/(?:last=(?P<days>\d+)/)?$',typingviews.DashboardData.as_view(),name='graph')
+    path('dashboard',typingviews.Dashboard.as_view(),name='streak'),
+    re_path(r'^getuserlog/(?:last=(?P<days>\d+)/)?$',typingviews.GraphData.as_view(),name='graph')
```
Comment (digi0ps) [inline typebackend/urls.py:13]: re_path is not actually needed here but it's fine!

File: typebackend/serializers.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (TheMightyRaider) [reply typebackend/serializers.py:26]: It doesn't have all the attribute, need to pass user attribute manually

File: typebackend/typingviews.py (hunk 5 above)
Comment (TheMightyRaider) [reply typebackend/typingviews.py:52]: A validation to check whether the front-end data is in the required format

File: typebackend/models.py [hunk 9]
```diff
@@ -24,3 +24,18 @@ class Paragraph(models.Model):
 
     def __str__(self):
         return self.taken_from
+
+class DashboardData(models.Model):
+    user=models.ForeignKey(settings.AUTH_USER_MODEL,on_delete=models.CASCADE,null=True)
```
Comment (digi0ps) [inline typebackend/models.py:29]: Can we make it OneToOne cause each user should be having one dashboard data object only.

Comment (digi0ps) [reply typebackend/models.py:29]: Why null=True?

File: typebackend/signals.py [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+
+    try:
+        last_typed=PractiseLog.objects.filter(user=user).latest('taken_at').taken_at
+    except PractiseLog.DoesNotExist:
+        # When its a new user, the first if condition has to be followed. Hence setting last_typed to previous_day
+        last_typed=new_entry-datetime.timedelta(1)
+
+    # Finding the difference in days between the last log and the new log.
+    days_between_recent_and_lastlog=(new_entry-last_typed).days
+    data=DashboardData.objects.get(user=user)
```
Comment (digi0ps) [inline typebackend/signals.py:23]: This won't be needed if dashboardData becomes one to one.

File: typebackend/signals.py [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+    if days_between_recent_and_lastlog==1:
+        data.streak=data.streak+1
+        data.longest_streak=data.streak if data.streak>data.longest_streak else data.longest_streak
+    # If the difference is greater, then reset the counter
+    if days_between_recent_and_lastlog>1:
+        data.inactive_days=data.inactive_days+(days_between_recent_and_lastlog-1)
+        data.streak=1
+        data.total_streak=data.total_streak+1
+
+    para_typed=DashboardData.objects.filter(user=user).annotate(total=F('arcade')+F('practise')+F('race'))[0].total+1
```
Comment (digi0ps) [inline typebackend/signals.py:40]: Same here. Filter won't be needed. So if we make it one to one we won't be making any extra queries.

File: typebackend/serializers.py [hunk 12]
```diff
@@ -18,13 +18,8 @@ def create(self,validated_data):
...
     class Meta:
         model=PractiseLog
-        fields=["para","wpm","taken_at","correct_words","wrong_words","total_words","accuracy"]
-        extra_kwargs = {'para': {'write_only': True}}
-
-    def create(self,validated_data):
-        user=self.context['request'].user
-        result=user.practiselog_set.create(para=validated_data['para'],wpm=validated_data['wpm'],correct_words=validated_data['correct_words'],wrong_words=validated_data['wrong_words'],total_words=validated_data['total_words'],accuracy=validated_data['accuracy'],taken_at=validated_data['taken_at'])
-        return result
+        fields="__all__"
```
Comment (digi0ps) [inline typebackend/serializers.py:21]: Awesome.

File: typebackend/typingviews.py [hunk 13]
```diff
@@ -62,19 +66,19 @@ def post(self,request):
...
+class GraphData(APIView):
     permission_classes=[IsAuthenticated]
 
     def get(self,request,days=0):  
         date_typed_log={}
-
         userlog=PractiseLog.objects.filter(user=request.user,taken_at__gte=timezone.now()-datetime.timedelta(days=int(days)))
         log_serializer=PractiseLogSerializer(userlog,many=True)
 
         for data in log_serializer.data:
```
Comment (digi0ps) [inline typebackend/typingviews.py:78]: Remove empty line

**Pull request merged**